### ISSUE TITLE
Impossible to pass a JSON string Array to consul_kv value parameter

### ISSUE THREAD

**Issue by mattupstate** [ansible/ansible-modules-extras#751]
##### Issue Type:

Bug Report
##### Ansible Version:

2.0.0
##### Ansible Configuration:

No changes
##### Environment:

OS X -> Ubuntu Trusty
##### Summary:

It does not seem possible to pass a list serialized as a JSON Array to the `consul_kv` module's `value` parameter.
##### Steps To Reproduce:

Run a playbook that looks like the following:

``` yaml

---
- hosts: all

  # ...
  consul_alerts_receivers:
    - 'sysadmins@somewhere.com'
  # ...

  tasks:
    - name: configure | email | set consul keys
      consul_kv:
        key: consul-alerts/config/notifiers/email/{{ item.key }}
        value: "{{ item.value | to_json | string }}"
        host: "{{ consul_alerts_consul_host }}"
        port: "{{ consul_alerts_consul_port }}"
      when: item.value is defined
      with_items:
        - key: enabled
          value: "true"
        - key: cluster-name
          value: "{{ consul_alerts_email_cluster_name }}"
        - key: url
          value: "{{ consul_alerts_email_url }}"
        - key: port
          value: "{{ consul_alerts_email_port }}"
        - key: username
          value: "{{ consul_alerts_email_username }}"
        - key: password
          value: "{{ consul_alerts_email_password }}"
        - key: sender-email
          value: "{{ consul_alerts_email_sender_email }}"
        - key: sender-alias
          value: "{{ consul_alerts_email_sender_alias }}"
        - key: receivers
          value: "{{ consul_alerts_email_receivers }}"
        - key: template
          value: "{{ consul_alerts_email_template }}"
```
##### Expected Results:

A JSON string is inserted in the Consul KV store at the desired key space.
##### Actual Results:

Ansible output:

```
TASK: [chatid.consul-alerts | configure | email | set consul keys] ************
ok: [machine1] => (item={'key': 'enabled', 'value': 'true'}) => {"changed": false, "data": {"CreateIndex": 8, "Flags": 0, "Key": "consul-alerts/config/notifiers/email/enabled", "LockIndex": 0, "ModifyIndex": 25, "Value": "\"true\""}, "index": "25", "item": {"key": "enabled", "value": "true"}, "key": "consul-alerts/config/notifiers/email/enabled"}
ok: [machine1] => (item={'key': 'cluster-name', 'value': u'Consul Alerts'}) => {"changed": false, "data": {"CreateIndex": 9, "Flags": 0, "Key": "consul-alerts/config/notifiers/email/cluster-name", "LockIndex": 0, "ModifyIndex": 26, "Value": "\"Consul Alerts\""}, "index": "26", "item": {"key": "cluster-name", "value": "Consul Alerts"}, "key": "consul-alerts/config/notifiers/email/cluster-name"}
ok: [machine1] => (item={'key': 'url', 'value': u'localhost'}) => {"changed": false, "data": {"CreateIndex": 10, "Flags": 0, "Key": "consul-alerts/config/notifiers/email/url", "LockIndex": 0, "ModifyIndex": 27, "Value": "\"localhost\""}, "index": "27", "item": {"key": "url", "value": "localhost"}, "key": "consul-alerts/config/notifiers/email/url"}
ok: [machine1] => (item={'key': 'port', 'value': u'25'}) => {"changed": false, "data": {"CreateIndex": 11, "Flags": 0, "Key": "consul-alerts/config/notifiers/email/port", "LockIndex": 0, "ModifyIndex": 28, "Value": "\"25\""}, "index": "28", "item": {"key": "port", "value": "25"}, "key": "consul-alerts/config/notifiers/email/port"}
ok: [machine1] => (item={'key': 'username', 'value': u'None'}) => {"changed": false, "data": {"CreateIndex": 12, "Flags": 0, "Key": "consul-alerts/config/notifiers/email/username", "LockIndex": 0, "ModifyIndex": 29, "Value": "\"None\""}, "index": "29", "item": {"key": "username", "value": "None"}, "key": "consul-alerts/config/notifiers/email/username"}
ok: [machine1] => (item={'key': 'password', 'value': u'None'}) => {"changed": false, "data": {"CreateIndex": 13, "Flags": 0, "Key": "consul-alerts/config/notifiers/email/password", "LockIndex": 0, "ModifyIndex": 30, "Value": "\"None\""}, "index": "30", "item": {"key": "password", "value": "None"}, "key": "consul-alerts/config/notifiers/email/password"}
ok: [machine1] => (item={'key': 'sender-email', 'value': u'alerts@machine1'}) => {"changed": false, "data": {"CreateIndex": 14, "Flags": 0, "Key": "consul-alerts/config/notifiers/email/sender-email", "LockIndex": 0, "ModifyIndex": 31, "Value": "\"alerts@machine1\""}, "index": "31", "item": {"key": "sender-email", "value": "alerts@machine1"}, "key": "consul-alerts/config/notifiers/email/sender-email"}
ok: [machine1] => (item={'key': 'sender-alias', 'value': u'None'}) => {"changed": false, "data": {"CreateIndex": 15, "Flags": 0, "Key": "consul-alerts/config/notifiers/email/sender-alias", "LockIndex": 0, "ModifyIndex": 32, "Value": "\"None\""}, "index": "32", "item": {"key": "sender-alias", "value": "None"}, "key": "consul-alerts/config/notifiers/email/sender-alias"}
failed: [machine1] => (item={'key': 'receivers', 'value': ['wright@chatid.com']}) => {"failed": true, "item": {"key": "receivers", "value": ["wright@chatid.com"]}}
msg: value should be None or a string / binary data
ok: [machine1] => (item={'key': 'template', 'value': u'None'}) => {"changed": false, "data": {"CreateIndex": 16, "Flags": 0, "Key": "consul-alerts/config/notifiers/email/template", "LockIndex": 0, "ModifyIndex": 33, "Value": "\"None\""}, "index": "33", "item": {"key": "template", "value": "None"}, "key": "consul-alerts/config/notifiers/email/template"}

FATAL: all hosts have already failed -- aborting
```

**Comment by robynbergeron** [comment]
[module: clustering/consul_kv.py]

@mattupstate -- Thanks for submitting this bug report.

@sgargan -- thanks in advance for taking a look at this bug report and for your ongoing work in maintaining this module.

**Comment by ghost** [comment]
is there any update to this?

**Comment by djalexd** [comment]
I managed to work around this issue by performing an HTTP call, see this post http://johndstein-blog.logdown.com/posts/277489-ansible-uri-body-bug.

It looks like `| string` filter doesn't do anything, the value received by both ansible extended module consul_kv and consul lib is dictionary.

If I am not mistaking (pretty new to Ansible), the HTTP put performed by consul library does not account for dictionary values (however they can be written with some help, i.e. `body_format=json`)

**Comment by bcoca** [comment]
it should just work with `value: "{{ item.value | to_json }}"`, the |string is actually counterproductive in this case.

**Comment by lepht** [comment]
There seems to be an issue with the CI tests, but for those still experiencing this issue the fix is super simple -- see the patch in #2125

**Comment by torresdal** [comment]
I stumbled across this issue and found a temporary workarond. If you add a `null` value to your json, Python will bail out and the value will be successfully sent as a string. Here's a failing example:

``` json
{
    "key": "value", 
    "otherKey": "otherValue", 
    "parent": {
        "child": "childValue"
    }
}
```

But by adding a `null` value like this, it works:

``` json
{
    "_hack": null,
    "key": "value", 
    "otherKey": "otherValue", 
    "parent": {
        "child": "childValue"
    }
}
```

**Comment by ansibot** [comment]
@sgargan, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by oryband** [comment]
Ping? The consul_kv module is useless right now.

**Comment by ansibot** [comment]
@sgargan, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@sgargan, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@sgargan, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by cgeers** [comment]
fwiw, I am also affected by this issue

**Comment by ansibot** [comment]
@sgargan, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sgargan** [comment]
@oryband @cgeers, I've checked the following against the 2.1 and 2.2 branches and can't reproduce. If you're still experiencing the issue could you report what's failing for you and I'll take a look.

``` yml
- hosts: localhost
  connection: local
  gather_facts: false

  vars:
    - list_of_items:
         - one:
             - a
         - two:
             - b
         - three:
             - 'sysadmins@somewhere.com'

  tasks:
    - name: store nested list of items
      consul_kv: key=somekey value={{list_of_items | to_json}}
      register: stored_list

    - debug: var=stored_list
```

**Comment by ansibot** [comment]
@sgargan, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sgargan** [comment]
Marking as wontfix.
